### PR TITLE
Use external library for date handling code

### PR DESCRIPTION
**Pull request by chrismytton**
This uses the [date_range](https://github.com/everypolitician/date_range) for handling date overlapping. I've taken this opportunity to refactor things into a `Membership` class, which encapsulates the date overlapping logic.

Fixes https://github.com/everypolitician/combine_popolo_memberships/issues/8

### CODE REVIEW

File: lib/combine_popolo_memberships.rb [hunk 1]
```diff
@@ -1,30 +1,50 @@
...
+      o = date_range.overlap(other.date_range)
+      return unless o
+      {}.merge(to_h).merge(other.to_h).tap do |h|
+        h[:start_date] = o.start_date
+        h[:end_date] = o.end_date
+      end
+    end
+
+    def to_h
+      raw
```
Comment (tmtmtmtm) [inline lib/combine_popolo_memberships.rb]: A public method that does nothing but expose a private method implies that something odd is going on somewhere…

File: lib/combine_popolo_memberships.rb [hunk 2]
```diff
@@ -1,30 +1,50 @@
...
+    end
+
+    def date_range
+      @date_range ||= DateRange.new(start_date, end_date)
+    end
+
+    def overlap(other)
+      o = date_range.overlap(other.date_range)
+      return unless o
+      {}.merge(to_h).merge(other.to_h).tap do |h|
```
Comment (tmtmtmtm) [inline lib/combine_popolo_memberships.rb]: This is a quite confusing construction. Why not just `to_h.merge`?

File: test/membership_test.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
+require 'test_helper'
+
+class MembershipTest < Minitest::Test
+  def start_date
+    '2015-11-27'
+  end
+
+  def end_date
+    '2015-12-31'
+  end
```
Comment (tmtmtmtm) [inline test/membership_test.rb]: Each of these are only used in a single place, so it's a bit odd to create methods for them…

File: test/membership_test.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    membership = CombinePopoloMemberships::Membership.new(start_date: '', end_date: '')
+    assert_equal '0000-00-00', membership.start_date
+    assert_equal '9999-99-99', membership.end_date
+  end
+
+  def test_returns_original_membership
+    membership = CombinePopoloMemberships::Membership.new(start_date: '', end_date: '')
+    assert_equal({ start_date: '', end_date: '' }, membership.to_h)
+    membership2 = CombinePopoloMemberships::Membership.new(start_date: nil, end_date: nil)
+    assert_equal({ start_date: nil, end_date: nil }, membership2.to_h)
```
Comment (tmtmtmtm) [inline test/membership_test.rb]: I think it would be clearer if the `to_h` tests were included with each of earlier groups, rather than in a separate method of its own.

File: lib/combine_popolo_memberships.rb [hunk 5]
```diff
@@ -1,30 +1,50 @@
...
-      _data: [mem, term],
-      start_date: s == '0000-00-00' ? nil : s,
-      end_date:   e == '9999-99-99' ? nil : e
-    }
+  class Membership
+    def initialize(raw)
+      @raw = raw
+    end
+
+    def start_date
```
Comment (tmtmtmtm) [inline lib/combine_popolo_memberships.rb]: I fear that this class has a bit too much in its public interface. It looks like `start_date` and `end_date` are only there so you can access them in the test? If so, I'd suggest making them private, and changing the test to get at the through the range instead.

File: lib/combine_popolo_memberships.rb [hunk 6]
```diff
@@ -1,30 +1,50 @@
...
-
-    return unless mS < tE && mE > tS
-    (s, e) = [mS, mE, tS, tE].sort[1, 2]
-    {
-      _data: [mem, term],
-      start_date: s == '0000-00-00' ? nil : s,
-      end_date:   e == '9999-99-99' ? nil : e
-    }
+  class Membership
+    def initialize(raw)
```
Comment (tmtmtmtm) [inline lib/combine_popolo_memberships.rb]: `raw` is a slightly awkward name here, especially as there's no documentation. It's not immediately obvious what this actually *is*.

File: lib/combine_popolo_memberships.rb [hunk 7]
```diff
@@ -1,30 +1,50 @@
...
-    from_data.product(into_data).map { |a, b| overlap(a, b) }.compact.map do |h|
-      from, to = h.delete :_data
-      [from, to].each { |orig|
-        orig.each { |k, v| h[k] = v unless %i(id start_date end_date).include? k }
-      }
+    from_data.product(into_data).map do |from, to|
+      h = Membership.new(from).overlap(Membership.new(to))
+      next unless h
+      h.delete(:id)
       h.merge(from_name => from[:id], into_name => to[:id])
```
Comment (tmtmtmtm) [inline lib/combine_popolo_memberships.rb:45]: it seems a little odd to still be accessing data directly from `from` and `to` here, when you've extracted a `Membership` to encapsulate them.

**Review by tmtmtmtm — CHANGES_REQUESTED** [review]
This was quite hard to follow, which on the one hand is understandable as the underlying code already is, but I fear that some of this is in the uncanny valley of being either a bit too big of a step (rather than simply switching in the library, and then refactoring separately), or too small of one. 

There are definitely places here where I'd be a lot more comfortable if we had a more comprehensive test-suite. I wonder if it's worth bulking that out a bit more separately, with more real-world examples from some of the places we've been using it (e.g. the double-step combination we're doing with the Dutch Wikidata)?

**Comment by chrismytton** [comment]
> There are definitely places here where I'd be a lot more comfortable if we had a more comprehensive test-suite. I wonder if it's worth bulking that out a bit more separately, with more real-world examples from some of the places we've been using it (e.g. the double-step combination we're doing with the Dutch Wikidata)?

Yeah, I agree. I'll add some more tests for the existing code in a separate PR, then come back to this and do the smaller step of swapping in date-range as a separate change.

File: lib/combine_popolo_memberships.rb (hunk 6 above)
Comment (chrismytton) [reply lib/combine_popolo_memberships.rb]: Agreed. I've changed it to `membership_hash` in b2a8a50.

File: lib/combine_popolo_memberships.rb (hunk 5 above)
Comment (chrismytton) [reply lib/combine_popolo_memberships.rb]: Good idea. I've switched to getting the dates through `date_range` in the tests in bfc873d and then made `start_date` and `end_date` private in 4dd049d.

File: lib/combine_popolo_memberships.rb (hunk 2 above)
Comment (chrismytton) [reply lib/combine_popolo_memberships.rb]: Ha, oops. Fixed in b860618.

File: lib/combine_popolo_memberships.rb (hunk 1 above)
Comment (chrismytton) [reply lib/combine_popolo_memberships.rb]: Yeah, not sure what's best to do about this. For now I've just exposed `membership_hash` (the new name for `raw`) in the public interface and dropped this `#to_h` method. 7d39d37

File: lib/combine_popolo_memberships.rb (hunk 7 above)
Comment (chrismytton) [reply lib/combine_popolo_memberships.rb:45]: The reason I've left this code alone for now is we're going to change the interface anyway when we fix #7, so I don't want to rush into anything now and add a `Membership#id` method which we might not need for the new interface.

So I think it's best to leave this as-is for now, then we can refactor this code in a future PR.

File: test/membership_test.rb (hunk 3 above)
Comment (chrismytton) [reply test/membership_test.rb]: Yep, I think they're a hangover from where I coped them from #2. Have removed them in 609e56f.

File: test/membership_test.rb (hunk 4 above)
Comment (chrismytton) [reply test/membership_test.rb]: Ah yep, done in 167c66d.

**Comment by chrismytton** [comment]
> This was quite hard to follow, which on the one hand is understandable as the underlying code already is, but I fear that some of this is in the uncanny valley of being either a bit too big of a step (rather than simply switching in the library, and then refactoring separately), or too small of one.

I don't think this was too big a step. The problem with just switching in the library and then refactoring separately is that the library isn't a drop-in replacement. We first need to get rid of the awkwardness of passing the two memberships back in the `:_data` key from `.overlap` in order to merge them together.

> There are definitely places here where I'd be a lot more comfortable if we had a more comprehensive test-suite. I wonder if it's worth bulking that out a bit more separately, with more real-world examples from some of the places we've been using it (e.g. the double-step combination we're doing with the Dutch Wikidata)?

I started writing tests for some real world data, but quickly realised that I was testing the `CombinePopoloMemberships.combine` interface, which is going to change soon when we fix #7. So I don't want to write loads of tests, only for us to decide that the output hash should be merged differently. The other option was to test `CombinePopoloMemberships.overlap`, but that functionality is going to be replaced with a combination of [`date-range`](https://github.com/everypolitician/date_range), which has its own test suite, and the `Membership` class in this pull request.

So I think the best thing to do is to get this merged, then we can write some more comprehensive tests for the `Membership` class and update the interface as enhancements on top of this change.

File: lib/combine_popolo_memberships.rb (hunk 7 above)
Comment (tmtmtmtm) [reply lib/combine_popolo_memberships.rb:45]: I don't really follow your thinking here, I'm afraid. The `Membership` class is essentially a private implementation — it doesn't _have_ a public interface. The only reason it seems to be exposed at all is so that you can unit test it (which I'm not sure has a lot of value in its own right, and definitely doesn't if it's already making things this rigid.)

It's not clear to me that that the `Membership` class will ever be directly part of the public interface — it's certainly not obvious so far that it will, though perhaps you've already thought through, or even have WIP code for, the next step. (If that _is_ the plan, then it worries me slightly that `Membership` may not be the right abstraction here, as one of the common usage patterns here is for one of the data lists to be a list of terms, which aren't really memberships)

**Comment by tmtmtmtm** [comment]
> I started writing tests for some real world data, but quickly realised that I was testing the `CombinePopoloMemberships.combine` interface, which is going to change soon when we fix #7. So I don't want to write loads of tests, only for us to decide that the output hash should be merged differently. The other option was to test `CombinePopoloMemberships.overlap`, but that functionality is going to be replaced with a combination of [`date-range`](https://github.com/everypolitician/date_range), which has its own test suite, and the `Membership` class in this pull request.

I think you've talked yourself into a false dichotomy here. The tests should be about making sure that, given a series of data hashes, the code slices and dices and puts them back again correctly — not just testing specific methods. If we're going to be changing the public interface anyway, then that makes the tests *more* valuable, not less. If you want to minimise how many changes will be needed to the tests when that happens, you could make the tests be largely data-driven, so that the code that interacts directly with the class is encapsulated into one place. But even without that, this shouldn't really be a problem. Any changes to the interface are presumably going to be done in a backwards compatible manner (e.g. adding a new interface (and testing that); then making the old interface delegate transparently to that (with the existing regression tests as the safety net that that doesn't change anything), then (later) dropping the old interface.

> So I think the best thing to do is to get this merged, then we can write some more comprehensive tests for the `Membership` class and update the interface as enhancements on top of this change.

Even if the code worked I'd be quite hesitant to do that, but I tried swapping this in to [the New Zealand scraper](https://github.com/everypolitician-scrapers/new-zealand-parliament), and it breaks quite badly.

For example, when dealing with Jami-Lee Ross, the old version ended up with these data hashes:

```ruby
{:start_date=>"2011-03-05", :end_date=>"2011-10-20", :name=>"49th Parliament", :wikidata=>"Q11812120", :area=>"Botany", :party=>"National Party", :term=>"49", :id=>"51MP3101"}
{:start_date=>"2011-12-20", :end_date=>"2014-09-19", :name=>"50th Parliament", :wikidata=>"Q4120749", :area=>"Botany", :party=>"National Party", :term=>"50", :id=>"51MP3101"}
{:start_date=>"2014-10-20", :end_date=>nil, :name=>"51st Parliament", :wikidata=>"Q4640115", :area=>"Botany", :party=>"National Party", :term=>"51", :id=>"51MP3101"}
```

whereas the new code gives:

```ruby
{:name=>"49th Parliament", :start_date=>"2011-03-05", :end_date=>"2011-10-20", :wikidata=>"Q11812120", :area=>"Botany", :party=>"National Party", :term=>"49", :id=>"51MP3101"}
{:name=>"50th Parliament", :start_date=>"2011-12-20", :end_date=>"2014-09-19", :wikidata=>"Q4120749", :area=>"Botany", :party=>"National Party", :term=>"50", :id=>"51MP3101"}
{:name=>"51st Parliament", :start_date=>"2014-10-20", :end_date=>"9999-99-99", :wikidata=>"Q4640115", :area=>"Botany", :party=>"National Party", :term=>"51", :id=>"51MP3101"}
```
The extra `name` field isn't an issue in this specific case, as we [explicitly filter the fields we want](https://github.com/everypolitician-scrapers/new-zealand-parliament/blob/bbeb251/scraper.rb#L36-L37), but it presumably would be in other cases, and leaking out the '9999-99-99' end date is a problem.

**Comment by tmtmtmtm** [comment]
> I don't think this was too big a step. The problem with just switching in the library and then refactoring separately is that the library isn't a drop-in replacement. 

Even if it's not a drop-in replacement, I'm still concerned that this PR is an awkward half-way house. Conceptually it's meant to have no change in user-facing behaviour — i.e. it's a refactoring. Yes, we would like to make the interface more intuitive, but *this* PR isn't about that — this is meant to just be simplifying the implementation. But right now I'm not convinced that it's doing that — it's introducing a new half-public-half-private class, that doesn't really hide the data it's encapsulating and may have a misleading name that will cause even more confusion later. Even in its current form I'm far from sure that this is an improvement, and, of course, if it doesn't actually work right yet, it's not clear what the final form will end up as.

I'm worried that fixing all this will burn up even more time and effort, when this isn't really on the critical path to anything. Unless I've forgotten a key part of why we're doing this work, the library already works well enough as is to do what we need. Tidying things up would definitely be nice, but is it really something we can spend weeks or even days on right now? Or, if we're doing this to enable something in particular, then we might need to rethink what's actually needed to get to that point.
